### PR TITLE
Add LATEST InsertStrategy to complement EARLIEST

### LINKED ISSUE #7611: Feature Request: Add LATEST InsertStrategy to complement EARLIEST
## Is your feature request related to a use case or problem? Please explain

Yes. The current `cirq.InsertStrategy` enum includes `EARLIEST`, which provides a powerful way to schedule an operation as early as possible in the circuit, pushing it to the "left." However, there is no symmetric counterpart to schedule an operation as late as possible, i.e., pushing it to the "right."

This missing functionality is useful for various circuit optimization and scheduling tasks. For example, a user may want to delay a gate as much as possible to mitigate decoherence before a measurement or to align operations for a specific hardware layout. The absence of a `LATEST` strategy forces users to implement this logic manually, which is cumbersome and error-prone.

---
## Describe the solution you would prefer

I propose adding a new `InsertStrategy` called **`LATEST`**. This strategy would be the logical inverse of `EARLIEST`.

Its behavior should be:

* Scan **forward** from the given insert location.
* Find the last possible moment where the new operation can be placed without conflicting with any existing operations on the same qubits.
* The operation should never be inserted at an index *before* the originally specified insertion location.
* If the scan reaches the end of the circuit without finding any conflicts, the operation should be added to the last moment if compatible, or a new moment at the very end.

A potential docstring could be:

```python
InsertStrategy.LATEST = InsertStrategy(
    'LATEST',
    """
    Scans forward from the insert location until a moment with operations
    touching qubits affected by the operation to insert is found. The operation
    is added into the moment just before that conflicting location.

    If the scan reaches the end of the circuit without finding any conflicting
    operations, the operation is added into the last moment of the circuit
    if possible, otherwise in a new moment at the end.

    The operation is never added into moments before the initial insert
    location.
    """,
)
```

## How urgent is this for you? Is it blocking important work?
P2 – I'm blocked by it, but will work around it.  

## Describe alternatives/workarounds you've considered
The primary workaround is to manually implement the logic. This involves getting the circuit's moments, iterating forward from the desired insertion index, checking for qubit conflicts at each moment (circuit.operation_at(qubit, moment_index)), and keeping track of the last valid placement. This is tedious, inefficient for users, and requires re-implementing logic that is core to the circuit construction API.

## Provide additional information or context
Adding LATEST would create a satisfying symmetry with the existing EARLIEST strategy. It would make the InsertStrategy API more complete, intuitive, and powerful for users performing circuit scheduling.

### PR DESCRIPTION
**Pull request by codrut3**
Scans forward from a given insert location and inserts an operation into the latest possible moment. If no moment is available, inserts the operation in a new moment.

Fixes #7611

### CODE REVIEW

**Comment by codrut3** [comment]
I updated `circuit.insert` to support the LATEST strategy.

Handling a single operation is rather straightforward. However, `insert` allows inserting an operation tree, so I spent some time thinking the best way to do this:

* I reverse the order in which operations are inserted (starting from last). Otherwise for something like `[cirq.X(q[0]), cirq.Y(q[0])]` the result of inserting in direct order with LATEST would be `[cirq.Y(q[0]), cirq.X(q[0])]` which doesn't make sense to me.
* The `insert` method still returns the moment index after the latest inserted operation. If someone calls insert repeatedly for consecutive operations, using  this value as the new insert index makes sense.
* Moments or operations are never inserted before the starting index `k`. This is different from EARLIEST, where the first batch is inserted in moments up to `k`, but the next batch can potentially be put in moment `k+1`, the second in moment `k+2`, and so on. If I can't insert at `k`, I add a new moment (which will then be the moment at index `k`).

File: cirq-core/cirq/circuits/circuit.py [hunk 1]
```diff
@@ -2160,6 +2160,85 @@ def _can_add_op_at(self, moment_index: int, operation: cirq.Operation) -> bool:
...
+        Args:
+            op: Operation for which the latest moment that can accommodate it needs to be found.
+            start_moment_index: The starting point of the reverse search. Defaults to 0.
+
+        Returns:
+            Index of the latest matching moment. Returns `start_moment_index - 1` if no moment on
+            the right is available, or `len(self._moments)` if `start_moment_index` equals it.
+        """
+        if start_moment_index is None:
+            start_moment_index = 0
```
Comment (NoureldinYosri) [inline cirq-core/cirq/circuits/circuit.py]: why note set the default to zero ?

File: cirq-core/cirq/circuits/circuit.py [hunk 2]
```diff
@@ -2160,6 +2160,85 @@ def _can_add_op_at(self, moment_index: int, operation: cirq.Operation) -> bool:
...
+        op_control_keys = protocols.control_keys(op)
+        op_measurement_keys = protocols.measurement_key_objs(op)
+        op_qubits = op.qubits
+        k = start_moment_index
+        while k < len(self._moments):
+            moment = self._moments[k]
+            if moment.operates_on(op_qubits):
+                return k - 1
+            moment_measurement_keys = moment._measurement_key_objs_()
+            if (
```
Comment (NoureldinYosri) [inline cirq-core/cirq/circuits/circuit.py]: [super optional] nit: turn the `not a or not b or not c` into `not (a and b and c)`

File: cirq-core/cirq/circuits/circuit.py (hunk 1 above)
Comment (pavoljuhas) [reply cirq-core/cirq/circuits/circuit.py]: +1 Since `_latest_available_moment` is an internal method with only one in-class caller, let us make the `start_moment_index` argument mandatory.

File: cirq-core/cirq/circuits/circuit.py [hunk 3]
```diff
@@ -2160,6 +2160,85 @@ def _can_add_op_at(self, moment_index: int, operation: cirq.Operation) -> bool:
...
+                or not moment._control_keys_().isdisjoint(op_measurement_keys)
+            ):
+                return k - 1
+            k += 1
+        return k - 1
+
+    def _insert_latest(self, k: int, batches: list[list[_MOMENT_OR_OP]]) -> int:
+        """Inserts batches of moments or operations using LATEST strategy.
+
+        Batches are inserted into reverse order.
```
Comment (pavoljuhas) [inline cirq-core/cirq/circuits/circuit.py]:
```suggestion
        Batches are processed in reverse order.
```

File: cirq-core/cirq/circuits/circuit.py [hunk 4]
```diff
@@ -2160,6 +2160,85 @@ def _can_add_op_at(self, moment_index: int, operation: cirq.Operation) -> bool:
...
+            k: The index to insert the batches at.
+            batches: Moments or operations to insert.
+
+        Returns:
+            The insertion index that will place operations just after the
+            operations that were inserted by this method.
+        """
+        batches = batches[::-1]
+        max_latest_index = -1  # Maximum index of a changed moment
+        for batch in batches:
```
Comment (pavoljuhas) [inline cirq-core/cirq/circuits/circuit.py]: LGTM with a couple of suggestions:

1. the corner case `c.insert(index, [])` returns `index` with the EARLIEST strategy, but 0 with the LATEST.  Let us make this consistent, ie, return `index` for either strategy.  Please also add a test for this case.

2. I found it a bit tricky to follow the `_insert_latest` code with Moment and Operation blocks interspersed.  Would you mind applying the patch below which regroups them to (a) Moment-only and (b) Operation only blocks?

```patch
diff --git a/cirq-core/cirq/circuits/circuit.py b/cirq-core/cirq/circuits/circuit.py
index 74c08354..6ecfcb41 100644
--- a/cirq-core/cirq/circuits/circuit.py
+++ b/cirq-core/cirq/circuits/circuit.py
@@ -2214,5 +2214,4 @@ class Circuit(AbstractCircuit):
         """
-        batches = batches[::-1]
         max_latest_index = -1  # Maximum index of a changed moment
-        for batch in batches:
+        for batch in reversed(batches):
             for moment_or_op in batch:
@@ -2220,22 +2219,22 @@ class Circuit(AbstractCircuit):
                 if isinstance(moment_or_op, Moment):
-                    p = k
+                    self._moments.insert(k, moment_or_op)
+                    max_latest_index = max(k, max_latest_index + 1)
                 else:
+                    end_moment_index = len(self.moments)
                     p = self._latest_available_moment(moment_or_op, start_moment_index=k)
                     if p < k:
-                        self._moments.insert(k, Moment())
-                        # All later moments shift by 1, so increase the max index
-                        max_latest_index += 1
-                        p = k
-                # Place
-                if isinstance(moment_or_op, Moment):
-                    self._moments.insert(p, moment_or_op)
-                    # All later moments shift by 1, so increase the max index
-                    max_latest_index += 1
-                elif p == len(self._moments):
-                    self._moments.append(Moment(moment_or_op))
-                else:
-                    self._moments[p] = self._moments[p].with_operation(moment_or_op)
-                max_latest_index = max(p, max_latest_index)
-        self._mutated(preserve_placement_cache=False)
-        return max_latest_index + 1
+                        self._moments.insert(k, Moment.from_ops(moment_or_op))
+                        max_latest_index = max(k, max_latest_index + 1)
+                    elif p < end_moment_index:
+                        self._moments[p] = self._moments[p].with_operation(moment_or_op)
+                        max_latest_index = max(p, max_latest_index)
+                    else:
+                        assert p == end_moment_index
+                        self._moments.append(Moment.from_ops(moment_or_op))
+                        max_latest_index = end_moment_index
+        # handle returned position index for empty batch
+        pos = k if max_latest_index == -1 else max_latest_index + 1
+        if max_latest_index != -1:
+            self._mutated(preserve_placement_cache=False)
+        return pos
 

```

File: cirq-core/cirq/circuits/circuit.py (hunk 1 above)
Comment (codrut3) [reply cirq-core/cirq/circuits/circuit.py]: Done.

File: cirq-core/cirq/circuits/circuit.py (hunk 2 above)
Comment (codrut3) [reply cirq-core/cirq/circuits/circuit.py]: Done.

File: cirq-core/cirq/circuits/circuit.py (hunk 3 above)
Comment (codrut3) [reply cirq-core/cirq/circuits/circuit.py]: Done.

File: cirq-core/cirq/circuits/circuit.py (hunk 4 above)
Comment (codrut3) [reply cirq-core/cirq/circuits/circuit.py]: Thank you! I changed as suggested.
